src/main/java/com/squareup/javapoet/TypeVariableName.java

```diff
@@ -43,6 +44,7 @@ private TypeVariableName(String name, List<TypeName> bounds, List<AnnotationSpec
     super(annotations);
     this.name = checkNotNull(name, "name == null");
     this.bounds = bounds;
+    this.boundsHashCode = bounds.hashCode();
```

Comment:
Both, `TypesTest` and `TypesEclipseTest`, run into a cycle, when line 117 in `AbstractTypesTest.getTypeVariableTypeMirrorRecursive()` the statement `typeVariableName.bounds.toString()` is executed. But only when:
 * `ParameterizedTypeName.hashCode()` and `TypeVariableName.hashCode()` calculate the internal collection hash code *on the fly*
 * it only happens if run via Maven/javac - if started inside Eclipse, there's no stack overflow produced

See failure report:
```
	at com.squareup.javapoet.ParameterizedTypeName.hashCode(ParameterizedTypeName.java:64)
	at java.util.AbstractList.hashCode(AbstractList.java:541)
	at java.util.Collections$UnmodifiableList.hashCode(Collections.java:1307)
	at com.squareup.javapoet.TypeVariableName.hashCode(TypeVariableName.java:72)
	at java.util.AbstractList.hashCode(AbstractList.java:541)
	at java.util.Collections$UnmodifiableList.hashCode(Collections.java:1307)
	at com.squareup.javapoet.ParameterizedTypeName.hashCode(ParameterizedTypeName.java:64)
	at java.util.AbstractList.hashCode(AbstractList.java:541)
	at java.util.Collections$UnmodifiableList.hashCode(Collections.java:1307)
	at com.squareup.javapoet.ParameterizedTypeName.hashCode(ParameterizedTypeName.java:64)


Results :

Tests in error: 
  getTypeVariableTypeMirrorRecursive(com.squareup.javapoet.TypesEclipseTest)
  getTypeVariableTypeMirrorRecursive(com.squareup.javapoet.TypesTest)

Tests run: 237, Failures: 0, Errors: 2, Skipped: 0
```
I move the entire hash code calculation into the constructor. Both classes are immutable and marked as final.